packages/react/src/index.ts

```diff
@@ -15,64 +17,183 @@ import {
+const SupportsProxy = typeof Proxy === "function";
+
+const ProxyHandlers = {
+	/**
+	 * This is a function call trap for functional components.
+	 * When this is called, we know it means React did run 'Component()',
+	 * that means we can use any hooks here to setup our effect and store.
+	 *
+	 * With the native Proxy, all other calls such as access/setting to/of properties will
+	 * be forwarded to the target Component, so we don't need to copy the Component's
+	 * own or inherited properties.
+	 *
+	 * @see https://github.com/facebook/react/blob/2d80a0cd690bb5650b6c8a6c079a87b5dc42bd15/packages/react-reconciler/src/ReactFiberHooks.old.js#L460
+	 */
+	apply(Component: FunctionComponent, thisArg: any, argumentsList: any) {
+		const store = useMemo(createEffectStore, Empty);
+
+		useSyncExternalStore(store.subscribe, store.getSnapshot, store.getSnapshot);
+
+		const stop = store.updater._start();
+
+		try {
+			const children = Component.apply(thisArg, argumentsList);
+			return children;
+		} catch (e) {
+			// Re-throwing promises that'll be handled by suspense
+			// or an actual error.
+			throw e;
+		} finally {
+			// Stop effects in either case before return or throw,
+			// Otherwise the effect will leak.
+			stop();
 		}
-	}
-	// @ts-ignore-next-line
-	return createElement.apply(this, arguments);
+	},
 };
 
-/*
-// This breaks React's controlled components implementation
-function createPropUpdater(props: any, prop: string, signal: Signal) {
-	let ref = props.ref;
-	if (!ref) ref = props.ref = React.createRef();
-	effect(() => {
-		if (props) props[prop] = signal.value;
-		let el = ref.current;
-		if (!el) return; // unsubscribe
-		(el as any)[prop] = signal.value;
-	});
-	props = null;
+function ProxyFunctionalComponent(Component: FunctionComponent<any>) {
+	return ProxyInstance.get(Component) || WrapWithProxy(Component);
 }
-*/
+function WrapWithProxy(Component: FunctionComponent<any>) {
+	if (SupportsProxy) {
+		const ProxyComponent = new Proxy(Component, ProxyHandlers);
 
-let finishUpdate: (() => void) | undefined;
-const updaterForComponent = new WeakMap<() => void, Effect>();
+		ProxyInstance.set(Component, ProxyComponent);
+		ProxyInstance.set(ProxyComponent, ProxyComponent);
 
-function setCurrentUpdater(updater?: Effect) {
-	// end tracking for the current update:
-	if (finishUpdate) finishUpdate();
-	// start tracking the new update:
-	finishUpdate = updater && updater._start();
+		return ProxyComponent;
+	}
+
+	/**
+	 * Emulate a Proxy if environment doesn't support it.
+	 *
+	 * @TODO - unlike Proxy, it's not possible to access the type/Component's
+	 * static properties this way. Not sure if we want to copy all statics here.
+	 * Omitting this for now.
+	 *
+	 * @example - works with Proxy, doesn't with wrapped function.
+	 * ```
+	 * const el = <SomeFunctionalComponent />
+	 * el.type.someOwnOrInheritedProperty;
+	 * el.type.defaultProps;
+	 * ```
+	 */
+	const WrappedComponent = function () {
```

Comment:
Really appreciate that you went the extra mile and added support for the non-proxy browsers too 👍